finished location service

tyisme614 opened the pull request:
finished address --> coordinate converter via gaode map api.
returned value could contain multiple results that are wrapped in an array.

tyisme614 committed:
Merge pull request #1 from wuhan2020/master

update arch diagram

tyisme614 committed:
finished location service for converting readable address to coordinate via gaode map api

frank-zsy commented on config/config.default.ts:
```diff
@@ -37,6 +37,19 @@ export default (appInfo: EggAppInfo) => {
+    }
+
+    //Baidu Map API Key
+    config.baidu = {
```
不是高德的接口吗？为什么需要配置百度的 KEY？

tyisme614 replied:
本来打算两个接口都做的，因为不知道前端用到的地图是什么，如果是百度地图的话，还会涉及到火星坐标的转换问题。所以这里的配置信息是预留的

frank-zsy replied:
我觉得可以先删除百度的配置，功能和配置同步

tyisme614 replied:
好的，新版本里删掉百度配置

frank-zsy replied:
OK，删掉以后重新 push 一下吧

frank-zsy commented:
另外，如果用 VSCode 作为 IDE，可以装一下 ESLint 插件，代码格式上的问题导致 CI 不过，提交前本地 npm test 一下

tyisme614 committed:
1.added map object for caching historical response data
2.added function for retrieving cached location data.
the function would calculate the similarity between the given address and cached address.
if the calculated similarity value is larger than 0.9, then the cached data would be returned.
3.removed Baidu configuration from config.default.ts

tyisme614 committed:
optimized code

frank-zsy commented on app/service/location.ts:
```diff
@@ -0,0 +1,136 @@
+
+export default class LocationService extends Service {
+
+  private dataCache = new Map();
```
Egg 框架中的 service 是与 ctx 绑定的，并非全局单例，这种写法，每次调度任务时 ctx 重新生成，cache 就会丢失。如果要做全局单例缓存，需要用 addSingleton 去挂，参考 https://eggjs.org/api/Application.html#addSingleton 。

frank-zsy commented on test/app/service/location.test.ts:
```diff
@@ -0,0 +1,25 @@
+    // let addr3 = '武昌区积玉桥街和平大道716号恒大世纪广场2602室';
+    // let addr4 = '洪山区高新二路光谷云计算高新企业孵化中心2号楼19楼';
+    const addr5 = '朝阳区郎家园138号[3-3]2幢1层C座666室';
+    const result1 = await ctx.service.location.retrieveCoordinateViaGD(addr1);
```
这里因为使用的一直是同一个 ctx，所以缓存是有效的

tyisme614 replied:
在链接https://eggjs.org/zh-cn/advanced/plugin.html#%E5%85%A8%E5%B1%80%E5%AE%9E%E4%BE%8B%E6%8F%92%E4%BB%B6%E7%9A%84%E6%9C%80%E4%BD%B3%E5%AE%9E%E8%B7%B5
中，官方有个一个例子讲解 addSingleton的写法。但是在我们的项目中并没有看到app.js这个脚本，也就是主脚本。那如果我想写一个全局的插件的话，在哪里调用 addSingleton呢？

tyisme614 replied:
btw  还有一种解决方案就是用redis。不知道最终项目跑起来是不是单线程，如果是多线程，就会涉及到多个app实例，这样的话单例也会有问题吧。用redis会不会好一点。

frank-zsy replied:
是在插件自己目录的 app.js 中，Egg 框架中每个插件可以独立发布，带有自己的 package.json，然后在 plugin 中配置路径即可加载，我们的项目目前没有插件，你可以写一个通用的缓存插件给各种 service 来用。参考一下上面文档中比较详细的说明，按步骤做就可以，或者看一下 https://github.com/eggjs/egg-mysql/blob/master/lib/mysql.js

frank-zsy replied:
是的，如果用 Redis 就没问题了，但需要额外的 Redis 实例，也可以使用这种方式，我来协调资源

tyisme614 committed:
Merge branch 'master' into master

tyisme614 committed:
added baidu geocoder function
added baidu api key configuration in config file

tyisme614 committed:
Merge remote-tracking branch 'origin/master'

tyisme614 committed:
Merge branch 'master' into master

tyisme614 commented:
可视化那边的同学说前端使用了百度地图，希望我们能够提供百度坐标系的坐标值，所以还是把百度的接口加上了，建议最终的数据表里也增加两个坐标组字段，一个是百度坐标系(bd09ll)，一个是通用坐标系(gcj02ll)。高德接口获得的坐标就是后者。

frank-zsy commented:
如果我没记错，高德地图获取到的是火星坐标，可以直接用算法转换为百度坐标，无需重新获取

tyisme614 committed:
Merge remote-tracking branch 'upstream/master'

tyisme614 commented:
我之前试过网上的转换算法将其它坐标系转换为百度的bd09ll坐标，貌似转换的效果不太好，所以后来干脆直接调用百度自己的坐标转换接口了。

tyisme614 committed:
Merge branch 'master' of https://github.com/tyisme614/data-sync

tyisme614 committed:
Merge branch 'master' of https://github.com/wuhan2020/data-sync

tyisme614 committed:
added service utility
call function checkURLConnectivity to check the url connectivity

tyisme614 committed:
Merge branch 'master' into master

tyisme614 commented:
新增utility，该服务用于一些功能性的操作，先实现了URL连通性检查。

tyisme614 committed:
Merge branch 'master' of https://github.com/wuhan2020/data-sync

# Conflicts:
#	package.json

tyisme614 committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	package.json

tyisme614 committed:
Merge branch 'master' of https://github.com/wuhan2020/data-sync

# Conflicts:
#	package.json

frank-zsy commented:
> 新增utility，该服务用于一些功能性的操作，先实现了URL连通性检查。

和之前的 PR 内容耦合了，应该另开分支做的，你这样的话如果缓存没改好，URL 检测也没法 merge 了啊

tyisme614 committed:
Merge branch 'master' of https://github.com/wuhan2020/data-sync

# Conflicts:
#	package.json

tyisme614 commented:
嗯  我新开branch来维护utility吧

tyisme614 committed:
implemented plugin datacache as a singleton instance, but it cannot be referenced in Service or Controller classes

tyisme614 committed:
added package.json of plugin

tyisme614 committed:
Merge branch 'master' into master

tyisme614 committed:
removed code in router.ts, this file has been deprecated

tyisme614 committed:
Merge branch 'master' of https://github.com/wuhan2020/data-sync

tyisme614 committed:
feat: implemented table schema class of doctor_hotle_table

tyisme614 committed:
merge: merged with remote repository

tyisme614 committed:
feat: implemented cache plugin by using singleton mode. The plugin is included in project/lib(no need to install via npm)

tyisme614 committed:
fix: disabled tests

tyisme614 committed:
fix: disabled tests

tyisme614 committed:
fix: disabled tests

tyisme614 committed:
Merge branch 'master' into plugin-datacache

tyisme614 committed:
Merge pull request #2 from tyisme614/plugin-datacache

merge branch code to master repository

tyisme614 committed:
fix: resolved lint errors

tyisme614 committed:
Merge branch 'master' into master

tyisme614 committed:
Merge branch 'master' into master

tyisme614 committed:
Merge branch 'master' of https://github.com/wuhan2020/data-sync

tyisme614 committed:
feat: implemented new function for retrieving place info from redis cache by using address as key

tyisme614 committed:
fix: disabled egg-redis

tyisme614 committed:
fix: added plugin state check

tyisme614 committed:
fix: enabled egg-redis

tyisme614 committed:
fix: added redis info

tyisme614 commented:
增加redis读取缓存的函数

frank-zsy commented:
Location service has been merged in another PR.
